jetty-util/src/main/java/org/eclipse/jetty/util/PathUtil.java

```diff
@@ -0,0 +1,157 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+/**
+ * Collection of Path encoding / decoding utilities for URI, Http Headers, and HTML
+ */
+public class PathUtil
+{
+    // Simple hex array
+    private static final char[] HEXES = "0123456789ABCDEF".toCharArray();
+    // The first 127 characters and their encodings.
+    private static final String[] URI_ENCODED_CHARS;
+
+    static
+    {
+        URI_ENCODED_CHARS = new String[127];
+        Arrays.fill(URI_ENCODED_CHARS, null);
+
+        // URI Reserved Chars
+        String uriReservedGenDelims = ":?#[]@"; // intentionally missing "/"
+        String uriReservedSubDelims = "!$&'()*+,;=";
+        // Extra Reserved Chars (specified by Jetty)
+        String jettyReserved = "%\"<> \\^`{}|";
+
+        String reserved = uriReservedGenDelims + uriReservedSubDelims + jettyReserved;
```

Comment:
I like this array approach, but I think we should switch around how we create it to defining what characters will not be encoded rather than which characters will be encoded.

We should not encode `/` and the unreserved characters from the URI spec:
 - ALPHA: `A` - `Z` + `a` - `z`
 - DIGIT: `0` - `9`
 - `-` / `.` / `_` / `~`

I think we should encode everything else.